pkg/controller/kubeconfig.go

```diff
@@ -6,60 +6,129 @@ import (
+	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/util/rand"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 )
 
 const (
 	secretTypeBootstrapToken v1.SecretType = "bootstrap.kubernetes.io/token"
+	machineNameLabelKey                    = "machine.k8s.io/machine.name"
+	tokenIDKey                             = "token-id"
+	tokenSecretKey                         = "token-secret"
+	expirationKey                          = "expiration"
+	tokenFormatter                         = "%s.%s"
 )
 
+func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
+	token, err := c.createBootstrapToken(name)
+	if err != nil {
+		return nil, err
+	}
+
+	infoKubeconfig, err := c.kubeconfigProvider.GetKubeconfig()
+	if err != nil {
+		return nil, err
+	}
+
+	outConfig := infoKubeconfig.DeepCopy()
+
+	outConfig.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		"": {
+			Token: token,
+		},
+	}
+
+	return outConfig, nil
+}
+
 func (c *Controller) createBootstrapToken(name string) (string, error) {
+	existingSecret, err := c.getSecretIfExists(name)
+	if err != nil {
+		return "", err
+	}
+	if existingSecret != nil {
+		return c.updateSecretExpirationAndGetToken(existingSecret)
+	}
+
 	tokenID := rand.String(6)
 	tokenSecret := rand.String(16)
 
 	secret := v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Name:   fmt.Sprintf("bootstrap-token-%s", tokenID),
+			Labels: map[string]string{machineNameLabelKey: name},
 		},
 		Type: secretTypeBootstrapToken,
 		StringData: map[string]string{
 			"description":                    "bootstrap token for " + name,
-			"token-id":                       tokenID,
-			"token-secret":                   tokenSecret,
-			"expiration":                     metav1.Now().Add(24 * time.Hour).Format(time.RFC3339),
+			tokenIDKey:                       tokenID,
+			tokenSecretKey:                   tokenSecret,
+			expirationKey:                    metav1.Now().Add(24 * time.Hour).Format(time.RFC3339),
 			"usage-bootstrap-authentication": "true",
 			"usage-bootstrap-signing":        "true",
 			"auth-extra-groups":              "system:bootstrappers:machine-controller:default-node-token",
 		},
 	}
 
-	_, err := c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
+	_, err = c.kubeClient.CoreV1().Secrets(metav1.NamespaceSystem).Create(&secret)
 	if err != nil {
 		return "", err
 	}
 
-	return fmt.Sprintf("%s.%s", tokenID, tokenSecret), nil
+	return fmt.Sprintf(tokenFormatter, tokenID, tokenSecret), nil
 }
 
-func (c *Controller) createBootstrapKubeconfig(name string) (*clientcmdapi.Config, error) {
-	token, err := c.createBootstrapToken(name)
+func (c *Controller) updateSecretExpirationAndGetToken(secret *v1.Secret) (string, error) {
+	if secret.StringData == nil {
+		secret.StringData = map[string]string{}
+	}
+	secret.StringData[expirationKey] = metav1.Now().Add(24 * time.Hour).Format(time.RFC3339)
```

Comment:
Can we lower the expiration time to `1 hour ?`

Reply:
Yep, makes sense. I think the 24h originate from `kubeadm`, but there the use case is a user using the token to manually join one or many nodes which is obviously different from what we do